internal/jujuclient/dial.go

```diff
@@ -331,7 +340,22 @@ func (c *Connection) Context() context.Context {
 // The given parameters are used as URL query values
 // when making the initial HTTP request.
 func (c *Connection) ConnectStream(path string, attrs url.Values) (base.Stream, error) {
-	return nil, errors.E(errors.CodeNotImplemented)
+	const op = errors.Op("jujuclient.ConnectStream")
+	user, pass, err := c.dialer.ControllerCredentialsStore.GetControllerCredentials(c.ctx, c.ctl.Name)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	requestHeader := jujuhttp.BasicAuthHeader(names.NewUserTag(user).String(), pass)
+
+	modelTag, ok := c.ModelTag()
+	if !ok {
+		return nil, errors.E(op, "no model found")
+	}
+	conn, err := rpc.Dial(c.ctx, c.ctl, modelTag, path, requestHeader)
```

Comment:
I think a rpc.DialWithBasicAuth would be a nice wrapper instead, and another one called something like rpc.DialNoAuth (or better naming plz aha), then you can put request header into it.

Also just move line 348 above this if you don't choose to do this change , because we compute the requestheader then get the model tag, but we waste the request header.

Reply:
I like the idea for `rpc.DialWithBasicAuth` but I think what we could do is refactor the parameter list into an object, then you can manipulate that object before passing it to dial potentially. 

I'd like to leave this for a follow-up, I'll implement the second suggestion.

Reply:
Not a fan of a params object in this scenario, we have two different behaviour dials. Lets just have two dial? I would think wrapping a struct up in params is really more for large amounts of params.

Reply:
Will leave this up to a follow-up PR. But moved the code around so that the request header is not created unnecessarily.